### DISCUSSION TITLE
Looping OrderedDict in reverse order

### DISCUSSION THREAD

**Discussion by mishal** [micropython/micropython#11881] [Using MicroPython] [answered]
I have an OrderedList of items and I would like to loop the items in reverse order like:

```
from collections import OrderedDict as TaskDict

d = TaskDict()
d["a"] = "something A"
d["c"] = "something C"
d["b"] = "something B"

print("Normal:")
for k, v in d.items():
    print(f"k: {k}, v: {v}")

print("Reversed:")
# reversed loop
for k, v in reversed(d.items()):
    print(f"k: {k}, v: {v}")
```

This produces following running on micropython (1.20.0, linux build):

```
Normal:
k: a, v: something A
k: c, v: something C
k: b, v: something B
Reversed:
Traceback (most recent call last):
  File "./tests/test_task_dict.py", line 15, in <module>
TypeError: object of type 'dict_view' has no len()
```

The code works fine using python 3.9.2 /3.5.2/, so the expected output should be:

```
Normal:
k: a, v: something A
k: c, v: something C
k: b, v: something B
Reversed:
k: b, v: something B
k: c, v: something C
k: a, v: something A
```
Looking at python's 3.5.2 docs. The docs here: https://docs.python.org/3.5/library/collections.html#collections.OrderedDict say that the object must have either `__reversed__` method or support sequence protocol... see https://docs.python.org/3.5/library/functions.html#reversed

Is this a bug or am I missing something?  Thanks

**Comment by mishal** [comment]
According to docs https://docs.micropython.org/en/latest/genrst/builtin_types.html?highlight=dict_view#dict which propose a workaround to **explicitly convert keys to a set before using set operations**. 

I converted them to a list which can be reverted using:

```
for k in reversed(list(d.keys())):
    v = d[k]
    print(f"k: {k}, v: {v}")
```

**Reply by jimmo** [reply]
@mishal Yep, this is the best workaround.

**Comment by sosi-deadeye** [comment, marked as answer]
Micropython is developed for Microcontrollers and must be very memory efficient. The __reversed__ method is not implemented for `OrderedDict` nor for `dict`.


Possible inefficient solution:
```
from collections import OrderedDict


class ItemsView:
    """
    ItemsView to support reversed
    But this implementation is memory-inefficient
    """
    def __init__(self, parent):
        self.parent = parent
    
    def __reversed__(self):
        for key in reversed(list(self.parent)):
            yield key, self.parent[key]


class TaskDict(OrderedDict):
    def items(self):
        return ItemsView(self)


print(list(reversed(TaskDict(a=10, b=20, c=30).items())))
```

It's simpler to use a `list` for tasks instead of a `dict`.
Or use `collections.deque`. It has the methods `append` and `popleft`.

**Comment by mishal** [comment]
Thanks @sosi-deadeye .

I need a dictionary, so for now I will just use this, I will try to implement my code using lists when optimizing the app.

Also to note is that micropython's `dict` method `keys()` does not return the keys in the order they were added, this is python 3.7+ language specification - see https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects
